notes/radicle-p2p-hypercore-overview.md

```diff
@@ -0,0 +1,126 @@
+Radicle Peer-to-peer Using `hypercore`
+======================================
+
+This document provides a high-level overview of how to build a P2P peer system
+to distribute and collaborate on Radicle chains using the
+[`hypercore`][hypercore] and [`discovery-swarm`][discovery-swarm] libraries.
+
+[hypercore]: https://github.com/mafintosh/hypercore
+[discovery-swarm]: https://github.com/mafintosh/discovery-swarm
+
+
+Hypercore
+---------
+
+`hypercore` is a library to distribute cryptographically secure append-only logs
+called *feeds*. A feed consists of a list of *entries* which may be JSON objects
+or binary blobs. A feed is signed with a private key. The hash of the
+corresponding public key is called the *discovery key* and acts as an identifier
+for the feed.
+
+Any user who knows the location (for example the IP and port) of a feed can read
+any entry in the feed, verify the integrity of any entry, listen to new entries,
+and replicate the feed.
+
+We say a user *hosts a feed* if the user makes it available to the world. If a
+users hosts a feed the user is either the owner of that feed or she replicates
+it.
+
+`hypercore` supports feed subscriptions. A user can subscribe to a feed and will
+be notified if the owner publishes a new entry.
+
+Every feed a user owns or replicates is stored locally and updated automatically
+if a user goes online.
+
+`discovery-swarm` is a library to discover hosts for hypercore feeds and
+announce the hosting of a feed. It uses the Bittorrent DHT and centralised DNS
+for global discovery as well as Multicast DNS for local discovery.
+
+
+Single-writer Chains
+--------------------
+
+*Single-writer chains* are Radicle chains that only the owner of a specific
+private key can write to.
+
+We use hypercore feeds to represent single-writer chains. Each entry in the feed
+is a Radicle expression. Single-writer chains are identified by the discovery
+key of their feed.
+
+To read a Radicle chain a user needs the discovery key of a feed. With the
+discovery key the user uses hyperdiscovery to find someone who replicates the
+feed. (This may be the owner.) The user then reads all entries in the feed with
+hypercore and evaluates them with radicle.
+
+To write an expression to a Radicle chain a user publishes a new entry to the
+`hypercore` feed.
+
+
+Multi-writer Chains
+-------------------
+
+*Multi-writer chains* are Radicle chains that the owners of multiple distinct
+private keys can write to.
+
+We implement a multi-writer chain by using set of feeds owned by different
+users. Every entry in these feeds is a Radicle expression and a link to another
```

Comment:
What does it mean if Alice wants to reference one of Bob's entries in a feed of hers? Does she:
1.   Fork bob's feed ?
2.  Create an entry in her already existing feed that has 2 pointers (one to the previous entry in her own feed, and one to this entry in bob's feed that she wants to reference) ?

If 1: What is the hypercore "public key" of her fork of bob's feed in this sense? how does her fork get represented as hypercore feeds?
If 2: Does multiple pointers in a feed entry mean multiple states/environments that a radical expression references?

Reply:
I’m not sure what you mean by “reference”. As I see it an entry in a feed only every link to one other entry in another feed. This is the link to the parent entry which determines the whole chain of Radicle expressions to follow.